subiquity/controllers/installprogress.py

```diff
@@ -40,10 +49,168 @@ class InstallState:
+    #             $SNAP2_$REV2.snap
+    #             $SNAP2_$REV2.assert
+    #           ...
+    #
+    # UpdateSnapSeed turns this into:
+    #
+    # /var/lib/snapd/seed/
+    #         seed.yaml -- updated seed
+    #         snaps/
+    #           [snaps from base seed]
+    #           $SNAP1_$REV1.snap
+    #           $SNAP2_$REV2.snap
+    #           ...
+    #         assertions/
+    #           [assertions from base seed]
+    #           $SNAP1_$REV1.assert
+    #           $SNAP2_$REV2.assert
+    #           ...
+
+    def __init__(self, controller, tmp_dir, snap_name, channel):
+        self.controller = controller
+        self.this_snap_download_dir = os.path.join(tmp_dir, snap_name)
+        self.snap_name = snap_name
+        self.channel = channel
+
+    def start(self):
+        self.controller._install_event_start(_("downloading {}").format(self.snap_name))
+        os.mkdir(self.this_snap_download_dir)
+        self.proc = utils.start_command(
+            ['snap', 'download', '--channel='+self.channel, self.snap_name],
+            cwd=self.this_snap_download_dir)
+
+    def _bg_run(self):
+        self.proc.communicate()
+        if self.proc.returncode != 0:
+            raise subprocess.CalledProcessError(
+                self.proc.returncode, self.proc.args, output=self.proc.stdout, stderr=self.proc.stderr)
+
+    def end(self, observer, fut):
+        self.controller._install_event_finish()
+        fut.result()
+        observer.task_succeeded()
+
+
+class UpdateSnapSeed(BackgroundTask):
+
+    def __init__(self, controller, root):
+        self.controller = controller
+        self.seed_yaml = os.path.join(root, "var/lib/snapd/seed/seed.yaml")
+        self.tmp_dir = os.path.join(root, "var/lib/snapd/seed/tmp")
+        self.snap_dir = os.path.join(root, "var/lib/snapd/seed/snaps")
+        self.assertions_dir = os.path.join(root, "var/lib/snapd/seed/assertions")
+
+    def start(self):
+        self.controller._install_event_start(_("updating snap seed"))
+
+    def _bg_run(self):
+        # This doesn't really need to be in the background, but as we
+        # have the infrastructure already in place, we may as well.
+        with open(self.seed_yaml) as fp:
+            seed = yaml.safe_load(fp)
+
+        for snap_name in os.listdir(self.tmp_dir):
+            this_snap_download_dir = os.path.join(self.tmp_dir, snap_name)
+            [snap_path] = glob.glob(os.path.join(this_snap_download_dir, "*.snap"))
+            [assertion_path] = glob.glob(os.path.join(this_snap_download_dir, "*.assert"))
+
+            snap_file = os.path.basename(snap_path)
+            assertion_file = os.path.basename(assertion_path)
+            os.rename(snap_path, os.path.join(self.snap_dir, snap_file))
+            os.rename(assertion_path, os.path.join(self.assertions_dir, assertion_file))
+
+            # If this directory is not empty, something very
+            # unexpected has happened and we should fail.
```

Comment:
Shouldn't we assert that instead?  Do we want a single snap download to be fatal to the entire install?

Reply:
I'd rather fail than have mysterious things happen I think. OTOH, your question about "Do we want a single snap download to be fatal to the entire install?" is fair enough, we should probably retry at least once and let the user request more retries (more complicated control flow incoming :/)